cmd/cli/docker/docker_run.go

```diff
@@ -106,143 +95,88 @@ func newDockerRunCmd() *cobra.Command { //nolint:funlen
 	}
 
-	return printer.PrintJobExecutionLegacy(ctx, executingJob, cmd, opts.DownloadSettings, opts.RunTimeSettings, util.GetAPIClient(ctx))
-}
-
-// CreateJob creates a job object from the given command line arguments and options.
-func CreateJob(ctx context.Context, image string, parameters []string, opts *DockerRunOptions) (*model.Job, error) {
-	outputs, err := parse.JobOutputs(ctx, opts.SpecSettings.OutputVolumes)
-	if err != nil {
-		return nil, err
+	if len(resp.Warnings) > 0 {
+		helpers.PrintWarnings(cmd, resp.Warnings)
 	}
 
-	nodeSelectorRequirements, err := parse.NodeSelector(opts.SpecSettings.Selector)
-	if err != nil {
-		return nil, err
+	if err := printer.PrintJobExecution(ctx, resp.JobID, cmd, opts.RunTimeSettings, api); err != nil {
+		return fmt.Errorf("failed to print job execution: %w", err)
 	}
 
-	labels, err := parse.Labels(ctx, opts.SpecSettings.Labels)
+	return nil
+}
+
+func build(args []string, opts *DockerRunOptions) (*models.Job, error) {
+	image := args[0]
+	parameters := args[1:]
+	engineSpec, err := engine_docker.NewDockerEngineBuilder(image).
+		WithParameters(parameters...).
+		WithWorkingDirectory(opts.WorkingDirectory).
+		WithEntrypoint(opts.Entrypoint...).
+		WithEntrypoint(opts.EnvironmentVariables...).
+		Build()
 	if err != nil {
 		return nil, err
 	}
 
-	spec, err := legacy_job.MakeDockerSpec(
-		image, opts.WorkingDirectory, opts.Entrypoint, opts.SpecSettings.EnvVar, parameters,
-		legacy_job.WithResources(
-			opts.ResourceSettings.CPU,
-			opts.ResourceSettings.Memory,
-			opts.ResourceSettings.Disk,
-			opts.ResourceSettings.GPU,
-		),
-		legacy_job.WithNetwork(
-			opts.NetworkingSettings.Network,
-			opts.NetworkingSettings.Domains,
-		),
-		legacy_job.WithTimeout(opts.SpecSettings.Timeout),
-		legacy_job.WithInputs(opts.SpecSettings.Inputs.Values()...),
-		legacy_job.WithOutputs(outputs...),
-		legacy_job.WithAnnotations(labels...),
-		legacy_job.WithNodeSelector(nodeSelectorRequirements),
-		legacy_job.WithDeal(
-			opts.DealSettings.TargetingMode,
-			opts.DealSettings.Concurrency,
-		),
-	)
-
-	// Publisher is optional and we won't provide it if not specified
-	p := opts.SpecSettings.Publisher.Value()
-	if p != nil {
-		spec.Publisher = p.Type //nolint:staticcheck
-		spec.PublisherSpec = *p
-	}
-
+	job, err := helpers.BuildJobFromFlags(engineSpec, opts.JobSettings, opts.TaskSettings)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("building job spec: %w", err)
 	}
 
-	return &model.Job{
-		APIVersion: model.APIVersionLatest().String(),
-		Spec:       spec,
-	}, nil
-}
-
-// dockerImageContainsTag checks if the image contains a tag or a digest
-func dockerImageContainsTag(image string) bool {
-	if strings.Contains(image, ":") {
-		return true
-	}
-	if strings.Contains(image, "@") {
-		return true
+	// Normalize and validate the job spec
+	job.Normalize()
+	if err := job.ValidateSubmission(); err != nil {
+		return nil, fmt.Errorf("%s: %w", userstrings.JobSpecBad, err)
```

Comment:
is there a way to put job normalization, validation, submission, and printing of progress in one place shared between `docker run`, `wasm run` and `job run`?

Reply:
Yes, although I'd like to wait on unification across the the CLI's until we migrate wasm to the new API spec. That alright with you?

Reply:
yeah sound good